### PR TITLE
Adding support for OpenSSH ecdsa-sk & ed25519-sk public keys

### LINKED ISSUE #10604: cryptography.exceptions.UnsupportedAlgorithm: Unsupported key type: b'sk-ssh-ed25519@openssh.com' - YubiKey SSH public keys
Hi,

using:

  * Python 3.12.2
  * cryptography 42.0.5
  * cffi         1.16.0

parsing the public key of a yubikey U2F ssh public key fails

```python
import cryptography.hazmat.primitives.serialization

cryptography.hazmat.primitives.serialization.load_ssh_public_key(b"sk-ssh-ed25519@openssh.com AAAAGnNrLXNzaC1lZDI1NTE5QG9wZW5zc2guY29tAAAAIMwRvXyiJg+fzGNsKeTm7gArZwzrDtnXTZvp9GcaClTaAAAADHNzaDpZdWJpa2V5Mw==")
```

```
cryptography.exceptions.UnsupportedAlgorithm: Unsupported key type: b'sk-ssh-ed25519@openssh.com'
```

Due to the unknown key format 
https://github.com/pyca/cryptography/blob/090bdf06016737e6df713f432c8b0c9fe5f871c1/src/cryptography/hazmat/primitives/serialization/ssh.py#L575-L582

and -later on- additional/unparsed application strings at the end of the container:

when force feeding a new format …
```python
cryptography.hazmat.primitives.serialization.ssh._KEY_FORMATS[b"sk-ssh-ed25519@openssh.com"] = cryptography.hazmat.primitives.serialization.ssh._KEY_FORMATS[b"ssh-ed25519"]
```

```python
cryptography.hazmat.primitives.serialization.load_ssh_public_key(b"sk-ssh-ed25519@openssh.com AAAAGnNrLXNzaC1lZDI1NTE5QG9wZW5zc2guY29tAAAAIMwRvXyiJg+fzGNsKeTm7gArZwzrDtnXTZvp9GcaClTaAAAADHNzaDpZdWJpa2V5Mw==")
```
```
ssh.py", line 1033, in _load_ssh_public_identity
    _check_empty(rest)
  File "~/venv/lib/python3.10/site-packages/cryptography/hazmat/primitives/serialization/ssh.py", line 184, in _check_empty
    raise ValueError("Corrupt data: unparsed data")
```

https://github.com/openssh/openssh-portable/blob/e27f032aa8fcbae9b2e7c451baaf4b8ac6fa3d45/PROTOCOL.u2f#L81-L86
```
The format of a sk-ssh-ed25519@openssh.com public key is:

	string		"sk-ssh-ed25519@openssh.com"
	string		public key
	string		application (user-specified, but typically "ssh:")
```

### PR DESCRIPTION
**Pull request by commonism**
fixes #10604

### CODE REVIEW

**Comment by commonism** [comment]
currently not 100% coverage:
```
src/cryptography/hazmat/primitives/serialization/ssh.py          774      3    241      0    99%   582, 600, 619
```

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 1]
```diff
@@ -572,13 +576,58 @@ def encode_private(
...
+        string		application (user-specified, but typically "ssh:")
+    """
+
+    def load_public(
+        self, data: memoryview
+    ) -> tuple[ec.EllipticCurvePublicKey, memoryview]:
+        """Make Ed25519 public key from data."""
+        public_key, data = super().load_public(data)
+        application, data = self.load_application(data)
+        return public_key, data
```
Comment (commonism) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:619]: uncovered?

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 2]
```diff
@@ -572,13 +576,58 @@ def encode_private(
...
+        string		application (user-specified, but typically "ssh:")
+    """
+
+    def load_public(
+        self, data: memoryview
+    ) -> tuple[ed25519.Ed25519PublicKey, memoryview]:
+        """Make Ed25519 public key from data."""
+        public_key, data = super().load_public(data)
+        application, data = self.load_application(data)
+        return public_key, data
```
Comment (commonism) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:600]: uncovered?

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 3]
```diff
@@ -572,13 +576,58 @@ def encode_private(
         f_priv.put_sshstr(f_keypair)
 
 
+class _SSHFormatSK:
+    def load_application(self, data) -> tuple[memoryview, memoryview]:
+        application, data = _get_sshstr(data)
+        return application, data
```
Comment (commonism) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:582]: uncovered?

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 4]
```diff
@@ -572,13 +576,58 @@ def encode_private(
...
+
+
+class _SSHFormatSKEd25519(_SSHFormatEd25519, _SSHFormatSK):
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:585]: We pretty strictly avoid subclassing our code. A better approach would be to have a `load_application` function (no class), and then in `_SSHFormatSKEd25519.load_pulic` you can do `_SSHFormatEd25519().load_public()` followed by a call to it

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 5]
```diff
@@ -572,13 +576,58 @@ def encode_private(
         f_priv.put_sshstr(f_keypair)
 
 
+class _SSHFormatSK:
+    def load_application(self, data) -> tuple[memoryview, memoryview]:
+        application, data = _get_sshstr(data)
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:581]: Given that we don't use the application, what do you think about raising an exception if it's not `ssh://`. We can later loosen this if it's necessary.

File: tests/hazmat/primitives/test_ssh.py [hunk 6]
```diff
@@ -80,10 +84,14 @@ def test_load_ssh_public_key(self, key_file, cert_file, backend):
                 )
         else:
             public_key = load_ssh_public_key(pub_data, backend)
-            assert (
-                public_key.public_bytes(Encoding.OpenSSH, PublicFormat.OpenSSH)
-                == nocomment_data
-            )
+            if not key_file.startswith("sk-"):
+                # sk U2F keys do not serialize
```
Comment (alex) [inline tests/hazmat/primitives/test_ssh.py:88]:
```suggestion
                # SK keys do not round-trip
```

File: src/cryptography/hazmat/primitives/serialization/ssh.py (hunk 5 above)
Comment (commonism) [reply src/cryptography/hazmat/primitives/serialization/ssh.py]: done + tests

File: src/cryptography/hazmat/primitives/serialization/ssh.py (hunk 4 above)
Comment (commonism) [reply src/cryptography/hazmat/primitives/serialization/ssh.py]: done - assuming subclassing _SSHFormatEd25519 is okay.

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 7]
```diff
@@ -572,13 +576,65 @@ def encode_private(
...
+    application, data = _get_sshstr(data)
+    if not application.tobytes().startswith(b"ssh:"):
+        raise ValueError(
+            "U2F application string does not start with b'ssh:' "
+            f"({application})"
+        )
+    return application, data
+
+
+class _SSHFormatSKEd25519(_SSHFormatEd25519):
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:592]: You shouldn't need to subclass `_SSHFormatEd25519` either, you should just be able to compose with it in `load_public`

Comment (commonism) [reply src/cryptography/hazmat/primitives/serialization/ssh.py]: done

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 8]
```diff
@@ -572,13 +576,67 @@ def encode_private(
...
+        string		public key
+        string		application (user-specified, but typically "ssh:")
+    """
+
+    def load_public(
+        self, data: memoryview
+    ) -> tuple[ed25519.Ed25519PublicKey, memoryview]:
+        """Make Ed25519 public key from data."""
+        global _KEY_FORMATS
+        public_key, data = _KEY_FORMATS[_SSH_ED25519].load_public(data)
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:606]:
```suggestion
        public_key, data = _SSHFormatEd25519().load_public(data)
```

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 9]
```diff
@@ -64,6 +64,10 @@ def _bcrypt_kdf(
 _ECDSA_NISTP521 = b"ecdsa-sha2-nistp521"
 _CERT_SUFFIX = b"-cert-v01@openssh.com"
 
+# U2F application string suffixed pubkey
+_SK_SSH_ED25519 = b"sk-ssh-ed25519@openssh.com"
+_SK_SSH_ECDSA = b"sk-ecdsa-sha2-nistp256@openssh.com"
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:69]:
```suggestion
_SK_SSH_ECDSA_NISTP256 = b"sk-ecdsa-sha2-nistp256@openssh.com"
```

File: src/cryptography/hazmat/primitives/serialization/ssh.py (hunk 8 above)
Comment (commonism) [reply src/cryptography/hazmat/primitives/serialization/ssh.py]: okay

File: src/cryptography/hazmat/primitives/serialization/ssh.py [hunk 10]
```diff
@@ -572,13 +576,65 @@ def encode_private(
...
+
 _KEY_FORMATS = {
     _SSH_RSA: _SSHFormatRSA(),
     _SSH_DSA: _SSHFormatDSA(),
     _SSH_ED25519: _SSHFormatEd25519(),
     _ECDSA_NISTP256: _SSHFormatECDSA(b"nistp256", ec.SECP256R1()),
     _ECDSA_NISTP384: _SSHFormatECDSA(b"nistp384", ec.SECP384R1()),
     _ECDSA_NISTP521: _SSHFormatECDSA(b"nistp521", ec.SECP521R1()),
+    _SK_SSH_ED25519: _SSHFormatSKEd25519(),
+    _SK_SSH_ECDSA: _SSHFormatSKECDSA(),
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/ssh.py:637]:
```suggestion
    _SK_SSH_ECDSA_NISTP256: _SSHFormatSKECDSA(),
```

**Review by alex — APPROVED** [review]

**Pull request merged**